gobblin-data-management/src/main/java/gobblin/data/management/retention/dataset/MultiVersionCleanableDatasetBase.java

```diff
@@ -0,0 +1,294 @@
+        IS_DATASET_BLACKLISTED_DEFAULT)), log);
+  }
+
+  /**
+   * Constructor for {@link MultiVersionCleanableDatasetBase}.
+   * @param fs {@link org.apache.hadoop.fs.FileSystem} where files are located.
+   * @param properties {@link java.util.Properties} for object.
+   * @param simulate whether to simulate deletes.
+   * @param skipTrash if true, delete files and directories immediately.
+   * @param deleteEmptyDirectories if true, newly empty parent directories will be deleted.
+   * @param deleteAsOwner if true, all deletions will be executed as the owner of the file / directory.
+   * @param log logger to use.
+   * @param isDatasetBlacklisted if true, clean will be skipped for this dataset
+   *
+   * @throws IOException
+   */
+  public MultiVersionCleanableDatasetBase(FileSystem fs, Properties properties, boolean simulate, boolean skipTrash,
+      boolean deleteEmptyDirectories, boolean deleteAsOwner, boolean isDatasetBlacklisted, Logger log) throws IOException {
+    this.log = log;
+    this.fs = fs;
+    this.simulate = simulate;
+    this.skipTrash = skipTrash;
+    this.deleteEmptyDirectories = deleteEmptyDirectories;
+    Properties thisProperties = new Properties();
+    thisProperties.putAll(properties);
+    if (this.simulate) {
+      thisProperties.setProperty(TrashFactory.SIMULATE, Boolean.toString(true));
+    }
+    if (this.skipTrash) {
+      thisProperties.setProperty(TrashFactory.SKIP_TRASH, Boolean.toString(true));
+    }
+    this.trash = TrashFactory.createProxiedTrash(this.fs, thisProperties);
+    this.deleteAsOwner = deleteAsOwner;
+    this.isDatasetBlacklisted = isDatasetBlacklisted;
+  }
+
+  public MultiVersionCleanableDatasetBase(FileSystem fs, Properties properties, boolean simulate, boolean skipTrash,
+      boolean deleteEmptyDirectories, boolean deleteAsOwner, Logger log) throws IOException {
+    this(fs, properties, simulate, skipTrash, deleteEmptyDirectories, deleteAsOwner, Boolean
+        .parseBoolean(IS_DATASET_BLACKLISTED_DEFAULT), log);
+  }
+
+  /**
+   * Perform the cleanup of old / deprecated dataset versions. See {@link gobblin.data.management.retention.DatasetCleanerNew}
+   * javadoc for more information.
+   * @throws java.io.IOException
+   */
+  @Override
+  public void clean() throws IOException {
+
+    if (this.isDatasetBlacklisted) {
+      log.info("Dataset blacklisted. Cleanup skipped for " + datasetRoot());
+      return;
+    }
+
+    for (VersionFinderAndPolicy<T> versionFinderAndPolicy : getVersionFindersAndPolicies()) {
+      VersionSelectionPolicy<T> selectionPolicy = versionFinderAndPolicy.getVersionSelectionPolicy();
+      VersionFinder<? extends T> versionFinder = versionFinderAndPolicy.getVersionFinder();
+
+      if (!selectionPolicy.versionClass().isAssignableFrom(versionFinder.versionClass())) {
+        throw new IOException("Incompatible dataset version classes.");
+      }
+
+      this.log.info("Cleaning dataset " + this);
```

Comment:
Can you add more information here, about using which versionFinder/versionSelectionPolicy?